File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/extensions/runanywhere_storage.dart
// SPDX-License-Identifier: Apache-2.0
//
// runanywhere_storage.dart — storage + download helpers.
// Mirrors Swift `RunAnywhere+Storage.swift`.

import 'package:fixnum/fixnum.dart';
import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/generated/model_types.pb.dart';
import 'package:runanywhere/generated/storage_types.pb.dart';
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/dart_bridge_file_manager.dart';
import 'package:runanywhere/native/dart_bridge_model_registry.dart';
import 'package:runanywhere/native/dart_bridge_storage.dart';
import 'package:runanywhere/public/extensions/model_category_extensions.dart';

/// Static helpers for storage + low-level download + model registration.
///
/// Mirrors Swift `RunAnywhere+Storage.swift` — every public method in the
/// Swift extension has a corresponding entry point here.
class RunAnywhereStorage {
  RunAnywhereStorage._();

  // ===========================================================================
  // Model registration (Swift-parity URL / archive / multi-file overloads)
  // ===========================================================================

  /// Register a remote model with the in-memory model registry from a
  /// download URL. Delegates the full build-and-save flow to the commons
  /// single-call factory `rac_register_model_from_url_proto`, which derives
  /// the canonical id from the URL (`rac_model_generate_id`), defaults
  /// format/framework/category/context-length, infers the artifact type from
  /// the URL extension (archive vs single-file), overlays the caller-supplied
  /// capability fields, and persists through the registry save path.
  ///
  /// Mirrors Swift `RunAnywhere.registerModel(id:name:url:framework:modality:
  /// artifactType:memoryRequirement:supportsThinking:supportsLora:)`, which
  /// routes through the same commons factory (`rac_model_info_make_proto`).
  static Future<ModelInfo> registerModel({
    String? id,
    required String name,
    required String url,
    required InferenceFramework framework,
    ModelCategory modality = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    ModelArtifactType? artifactType,
    int? memoryRequirement,
    bool supportsThinking = false,
    bool supportsLora = false,
  }) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }

    final request = RegisterModelFromUrlRequest(
      url: url,
      name: name,
      framework: framework,
      category: modality,
      source: ModelSource.MODEL_SOURCE_REMOTE,
      supportsThinking: supportsThinking,
      supportsLora: supportsLora,
    );
    // Caller-supplied id always wins; otherwise commons derives it from the
    // URL (matching Swift's `generatedModelID(from:name:)`).
    if (id != null) {
      request.id = id;
    }
    // Explicit artifact-type override wins over commons' URL inference.
    if (artifactType != null) {
      request.artifactType = artifactType;
    }
    if (memoryRequirement != null) {
      request.memoryRequiredBytes = Int64(memoryRequirement);
    }
    // Intentionally NOT setting downloadSizeBytes from memoryRequirement: that
    // value gates the post-finalize download-size check, and the RAM estimate
    // is usually a round placeholder (e.g. 500 MB for a real 397 MB file),
    // which leaves is_downloaded=false forever. Leaving it unset lets commons
    // validate against the actual transfer — matches Kotlin's catalog.

    final model = await DartBridgeModelRegistry.instance.registerModelFromUrl(
      request,
    );
    if (model == null) {
      throw SDKException.internalError(
        'rac_register_model_from_url_proto failed for model "$name"',
      );
    }
    return model;
  }

  /// Register an archive-packaged model (tar.gz / tar.bz2 / tar.xz / zip)
  /// where the caller needs to specify the on-disk layout
  /// ([ArchiveStructure.ARCHIVE_STRUCTURE_DIRECTORY_BASED],
  /// [ArchiveStructure.ARCHIVE_STRUCTURE_NESTED_DIRECTORY], etc.) the
  /// URL-form [registerModel] cannot infer.
  ///
  /// Composes [registerModel] and then patches the resolved
  /// [ArchiveArtifact.structure] before re-saving through the registry.
  ///
  /// Mirrors Swift `RunAnywhere.registerModel(archive:structure:...)`.
  static Future<ModelInfo> registerArchiveModel({
    required String archiveUrl,
    required ArchiveStructure structure,
    String? id,
    required String name,
    required InferenceFramework framework,
    ModelCategory modality = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    ArchiveType? archiveType,
    int? memoryRequirement,
    bool supportsThinking = false,
    bool supportsLora = false,
  }) async {
    // Map an explicit caller archive type to its artifact-type override.
    // When the caller passes none, leave the override unset so commons
    // infers the archive type from the URL extension (mirroring Swift's
    // `ArchiveType.from(url:)`, which lives behind
    // `rac_model_info_make_proto`).
    final ModelArtifactType? resolvedArtifactType;
    if (archiveType == null) {
      resolvedArtifactType = null;
    } else {
      switch (archiveType) {
        case ArchiveType.ARCHIVE_TYPE_ZIP:
          resolvedArtifactType =
              ModelArtifactType.MODEL_ARTIFACT_TYPE_ZIP_ARCHIVE;
          break;
        case ArchiveType.ARCHIVE_TYPE_TAR_GZ:
          resolvedArtifactType =
              ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_GZ_ARCHIVE;
          break;
        case ArchiveType.ARCHIVE_TYPE_TAR_BZ2:
          resolvedArtifactType =
              ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_BZ2_ARCHIVE;
          break;
        case ArchiveType.ARCHIVE_TYPE_TAR_XZ:
          resolvedArtifactType =
              ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_XZ_ARCHIVE;
          break;
        default:
          resolvedArtifactType = ModelArtifactType.MODEL_ARTIFACT_TYPE_ARCHIVE;
      }
    }

    var model = await registerModel(
      id: id,
      name: name,
      url: archiveUrl,
      framework: framework,
      modality: modality,
      artifactType: resolvedArtifactType,
      memoryRequirement: memoryRequirement,
      supportsThinking: supportsThinking,
      supportsLora: supportsLora,
    );

    // Preserve the caller-specified layout on the archive artifact. Commons
    // infers the archive type from the URL; the nested/directory layout can
    // only come from the caller — patch and re-persist (mirroring Swift's
    // archive overload, which always carries an archive artifact with the
    // caller's structure).
    final patchedArchive =
        (model.hasArchive() ? model.archive.deepCopy() : ArchiveArtifact())
          ..structure = structure;
    if (archiveType != null) {
      patchedArchive.type = archiveType;
    }
    model = model.deepCopy()
      ..archive = patchedArchive
      ..updatedAtUnixMs = Int64(DateTime.now().millisecondsSinceEpoch);
    await DartBridgeModelRegistry.instance.saveProtoModel(model);
    return model;
  }

  /// Register a multi-file model (e.g., VLMs with a separate `mmproj`,
  /// MiniLM embedding with `vocab.txt`) through the canonical commons
  /// factory (`rac_register_multi_file_model_proto`) — no URL is involved at
  /// the model level because each [ModelFileDescriptor] carries its own URL.
  ///
  /// Mirrors Swift `RunAnywhere.registerModel(multiFile:id:name:framework:
  /// modality:memoryRequirement:contextLength:supportsThinking:source:)`.
  static Future<ModelInfo> registerMultiFileModel({
    required List<ModelFileDescriptor> files,
    required String id,
    required String name,
    required InferenceFramework framework,
    ModelCategory modality = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    int? memoryRequirement,
    int? contextLength,
    bool supportsThinking = false,
    ModelSource source = ModelSource.MODEL_SOURCE_REMOTE,
  }) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }

    // Canonical commons factory: rac_register_multi_file_model_proto builds
    // the MultiFileArtifact ModelInfo (descriptors carry url/filename/size/
    // checksum/role) and persists it with merge-on-reseed semantics.
    final request = RegisterMultiFileModelRequest(
      id: id,
      name: name,
      framework: framework,
      category: modality,
      supportsThinking: supportsThinking,
      source: source,
      files: files,
    );
    if (memoryRequirement != null) {
      request.memoryRequiredBytes = Int64(memoryRequirement);
    }
    // See registerModel: downloadSizeBytes is intentionally left unset so the
    // post-finalize size guard validates against the actual transfer rather
    // than the RAM-estimate placeholder.
    final resolvedContextLength =
        contextLength ?? (modality.requiresContextLength ? 2048 : null);
    if (resolvedContextLength != null) {
      request.contextLength = resolvedContextLength;
    }

    final model =
        await DartBridgeModelRegistry.instance.registerMultiFileModel(request);
    if (model == null) {
      throw SDKException.internalError(
        'rac_register_multi_file_model_proto failed for model "$name"',
      );
    }
    return model;
  }

  /// Import a stable, platform-normalized local model path into the
  /// generated registry. Also the public local-import entry point for
  /// file-picker / bookmark flows after the platform has handled sandbox
  /// access.
  ///
  /// Mirrors Swift `RunAnywhere.importModel(_:)`. Backed by
  /// `rac_model_registry_import_proto`.
  static Future<ModelImportResult> importModel(
    ModelImportRequest request,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeModelRegistry.instance.importModel(request);
  }

  // ===========================================================================
  // Storage availability (existing Flutter-specific helpers)
  // ===========================================================================

  /// Clear the SDK's Temp directory. Mirrors Swift's `cleanTempFiles()`.
  static Future<void> cleanTempFiles() async {
    if (!DartBridgeFileManager.clearTemp()) {
      throw SDKException.storageError('Failed to clean temp files');
    }
  }

  /// Execute a generated-proto storage delete request.
  ///
  /// Mirrors Swift `RunAnywhere.deleteStorage(_:)`; callers choose the typed
  /// policy flags (`deleteFiles`, `clearRegistryPaths`, `unloadIfLoaded`,
  /// `allowPlatformDelete`) while commons owns the actual plan/execution.
  static Future<StorageDeleteResult> deleteStorage(
    StorageDeleteRequest request,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeStorage.instance.deleteProto(request);
  }

  /// Delete one downloaded model end-to-end: unload it if loaded, remove its
  /// files through the platform adapter, and clear its registry path so the
  /// entry returns to registered-not-downloaded (re-downloadable).
  /// Convenience over [deleteStorage] with the canonical flag set — mirrors
  /// Swift `RunAnywhere.deleteModel(_:)`.
  static Future<StorageDeleteResult> deleteModel(String modelId) {
    return deleteStorage(
      StorageDeleteRequest(
        modelIds: [modelId],
        deleteFiles: true,
        clearRegistryPaths: true,
        unloadIfLoaded: true,
        allowPlatformDelete: true,
      ),
    );
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/extensions/Storage/RunAnywhereStorage.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for storage operations.
 *
 * Mirrors Swift `RunAnywhere+Storage.swift` exactly:
 *   - `registerModel(...)` URL / archive / multi-file overloads (Swift parity)
 *   - `importModel(request)` local-import entry point
 *   - `getStorageInfo(request)` (replaces the legacy `storageInfo()` accessor)
 *   - `deleteStorage(request)` executing or dry-running deletion
 *   - `clearCache()` / `cleanTempFiles()` forwarding to the FileManager bridge
 */

package com.runanywhere.sdk.public.extensions

import ai.runanywhere.proto.v1.ArchiveStructure
import ai.runanywhere.proto.v1.ArchiveType
import ai.runanywhere.proto.v1.InferenceFramework
import ai.runanywhere.proto.v1.ModelArtifactType
import ai.runanywhere.proto.v1.ModelCategory
import ai.runanywhere.proto.v1.ModelFileDescriptor
import ai.runanywhere.proto.v1.ModelImportRequest
import ai.runanywhere.proto.v1.ModelImportResult
import ai.runanywhere.proto.v1.ModelSource
import ai.runanywhere.proto.v1.RegisterModelFromUrlRequest
import ai.runanywhere.proto.v1.RegisterMultiFileModelRequest
import ai.runanywhere.proto.v1.StorageDeleteRequest
import ai.runanywhere.proto.v1.StorageDeleteResult
import ai.runanywhere.proto.v1.StorageInfoRequest
import ai.runanywhere.proto.v1.StorageInfoResult
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeFileManager
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeModelRegistry
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeStorage
import com.runanywhere.sdk.foundation.errors.SDKException
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.Models.archiveArtifact
import com.runanywhere.sdk.public.extensions.Models.setArchiveArtifact
import com.runanywhere.sdk.public.types.RAModelInfo
import com.runanywhere.sdk.utils.getCurrentTimeMillis

// MARK: - Model Registration

// MARK: - Model Import

// MARK: - Storage Information

// MARK: - Storage Deletion

// MARK: - Cache and Temp

private fun requireStorageInitialized(sdk: RunAnywhere) {
    if (!sdk.isInitialized) {
        throw SDKException.notInitialized("SDK not initialized")
    }
}

// MARK: - Model Registration

suspend fun RunAnywhere.registerModel(
    id: String? = null,
    name: String,
    url: String,
    framework: InferenceFramework,
    modality: ModelCategory = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    artifactType: ModelArtifactType? = null,
    memoryRequirement: Long? = null,
    supportsThinking: Boolean = false,
    supportsLora: Boolean = false,
): RAModelInfo {
    requireStorageInitialized(this)

    // Canonical commons factory: rac_register_model_from_url_proto derives
    // id/name/format/artifact, resolves hf.co/org/repo[:quant] refs (quant
    // selection, mmproj pairing, sharded GGUF sets, checksums), and preserves
    // prior download state when a catalog re-seeds on launch.
    val request =
        RegisterModelFromUrlRequest(
            url = url,
            name = name,
            framework = framework,
            category = modality,
            source = ModelSource.MODEL_SOURCE_REMOTE,
            id = id,
            memory_required_bytes = memoryRequirement,
            supports_thinking = if (supportsThinking) true else null,
            supports_lora = if (supportsLora) true else null,
            artifact_type = artifactType,
        )

    val saved =
        CppBridgeModelRegistry.registerModelFromUrl(request)
            ?: throw SDKException.storage(
                "Native registry proto ABI unavailable for registerModelFromUrl",
            )
    return saved
}

suspend fun RunAnywhere.registerModel(
    archiveUrl: String,
    structure: ArchiveStructure,
    id: String? = null,
    name: String,
    framework: InferenceFramework,
    modality: ModelCategory = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    archiveType: ArchiveType? = null,
    memoryRequirement: Long? = null,
    supportsThinking: Boolean = false,
    supportsLora: Boolean = false,
): RAModelInfo {
    val resolvedArtifactType: ModelArtifactType? =
        archiveType?.let { type ->
            when (type) {
                ArchiveType.ARCHIVE_TYPE_ZIP -> ModelArtifactType.MODEL_ARTIFACT_TYPE_ZIP_ARCHIVE
                ArchiveType.ARCHIVE_TYPE_TAR_GZ -> ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_GZ_ARCHIVE
                ArchiveType.ARCHIVE_TYPE_TAR_BZ2 -> ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_BZ2_ARCHIVE
                ArchiveType.ARCHIVE_TYPE_TAR_XZ -> ModelArtifactType.MODEL_ARTIFACT_TYPE_TAR_XZ_ARCHIVE
                ArchiveType.ARCHIVE_TYPE_UNSPECIFIED -> ModelArtifactType.MODEL_ARTIFACT_TYPE_ARCHIVE
            }
        }

    var model =
        registerModel(
            id = id,
            name = name,
            url = archiveUrl,
            framework = framework,
            modality = modality,
            artifactType = resolvedArtifactType,
            memoryRequirement = memoryRequirement,
            supportsThinking = supportsThinking,
            supportsLora = supportsLora,
        )

    // Preserve the structure on the archive artifact. The URL-form inferred
    // artifact only captures the archive type, not the nested/directory
    // layout, so patch it here and re-persist through the registry's proto
    // save path (mirroring Swift's archive overload).
    val archive = model.archiveArtifact ?: return model
    model =
        model
            .setArchiveArtifact(archive.copy(structure = structure))
            .copy(updated_at_unix_ms = getCurrentTimeMillis())
    CppBridgeModelRegistry.save(model)
    return model
}

suspend fun RunAnywhere.registerModel(
    multiFile: List<ModelFileDescriptor>,
    id: String,
    name: String,
    framework: InferenceFramework,
    modality: ModelCategory = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    memoryRequirement: Long? = null,
    contextLength: Int? = null,
    supportsThinking: Boolean = false,
    source: ModelSource = ModelSource.MODEL_SOURCE_REMOTE,
): RAModelInfo {
    requireStorageInitialized(this)

    // Canonical commons factory: rac_register_multi_file_model_proto builds
    // the MultiFileArtifact ModelInfo (descriptors carry url/filename/
    // size/checksum/role) and persists it with merge-on-reseed semantics.
    val request =
        RegisterMultiFileModelRequest(
            id = id,
            name = name,
            framework = framework,
            category = modality,
            memory_required_bytes = memoryRequirement,
            context_length = contextLength,
            supports_thinking = if (supportsThinking) true else null,
            source = source,
            files = multiFile,
        )

    val saved =
        CppBridgeModelRegistry.registerMultiFileModel(request)
            ?: throw SDKException.storage(
                "Native registry proto ABI unavailable for registerMultiFileModel",
            )
    return saved
}

// MARK: - Model Import

suspend fun RunAnywhere.importModel(request: ModelImportRequest): ModelImportResult {
    requireStorageInitialized(this)
    ensureServicesReady()
    return CppBridgeModelRegistry.importModel(request)
}

// MARK: - Storage Information

suspend fun RunAnywhere.getStorageInfo(request: StorageInfoRequest = StorageInfoRequest()): StorageInfoResult {
    requireStorageInitialized(this)
    ensureServicesReady()
    return CppBridgeStorage.info(request)
        ?: throw SDKException.storage("Native storage info proto API unavailable")
}

suspend fun RunAnywhere.deleteStorage(request: StorageDeleteRequest): StorageDeleteResult {
    requireStorageInitialized(this)
    ensureServicesReady()
    return CppBridgeStorage.delete(request)
        ?: throw SDKException.storage("Native storage delete proto API unavailable")
}

/**
 * Delete one downloaded model end-to-end: unload it if loaded, remove its
 * files through the platform adapter, and clear its registry path so the
 * entry returns to registered-not-downloaded (re-downloadable). Convenience
 * over [deleteStorage] with the canonical flag set — mirrors Swift
 * `RunAnywhere.deleteModel(_:)`.
 */
suspend fun RunAnywhere.deleteModel(modelId: String): StorageDeleteResult =
    deleteStorage(
        StorageDeleteRequest(
            model_ids = listOf(modelId),
            delete_files = true,
            clear_registry_paths = true,
            unload_if_loaded = true,
            allow_platform_delete = true,
        ),
    )

suspend fun RunAnywhere.clearCache() {
    requireStorageInitialized(this)
    ensureServicesReady()
    if (!CppBridgeFileManager.clearCache()) {
        throw SDKException.storage("Failed to clear cache")
    }
}

suspend fun RunAnywhere.cleanTempFiles() {
    requireStorageInitialized(this)
    ensureServicesReady()
    if (!CppBridgeFileManager.clearTemp()) {
        throw SDKException.storage("Failed to clean temp files")
    }
}

// MARK: - Helpers
